src/bindings/gpio/gpio_controller.rs

```diff
@@ -0,0 +1,223 @@
+const GPIO_CONTROLLER_IS_TAKEN: AtomicBool = AtomicBool::new(false);
+
+/// Implementation of a memory map (dev/gpiomem) gpio controller, which is not thread safe.
+/// I.e. the use of this should only come from on thread, if no care to make it thread safe while using it is taken.
+///
+/// This structure allow you to controll a output gpio, i.e. the gpio let the current pass or not
+///
+/// ### Examples
+///         
+/// let mut gpio_controller = GpioController::get_the_gpio_controller();
+///
+/// let mut gpio_output_pin = GpioOutputPin::new(GpioPin::new(& GpioPinAvailable::Gpio0), & gpio_controller);
+///
+/// gpio_controller.set_high(& gpio_outputpin);
+///
+pub struct GpioController {
+    /// A u32 pointer in C
+    // as registers or of size x pointer should be ux
+    mem_ptr: *mut u32,
+}
+
+impl GpioController {
+    /// Give back the singleton [GpioController] or panic
+    ///
+    /// The function give you back the singleton [GpioController] if it is not present in the process that call it.
+    ///
+    /// ## Panics
+    /// The function panic if you call it when there is already a instance of [GpioController] in the process you are in.
+    /// When the lifecyle of the GpioController is finished, you can get a [GpioController] again.
+    ///
+    pub fn get_the_gpio_controller() -> GpioController {
+        // Will panic if we try to get multiple instance of GpioController
+        GPIO_CONTROLLER_IS_TAKEN
+            .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
+            .expect("Try to obtain multiple GpioController");
+        GpioController::new()
+    }
+
+    /// Instantiate a [GpioController]
+    ///
+    /// ## Panics
+    /// see [map_devgpiomem] panics
+    ///
+    fn new() -> GpioController {
+        let mem_ptr = Self::map_devgpiomem();
+        GpioController { mem_ptr: mem_ptr }
+    }
+
+    /// Return the start user space address of the gpio interface registers.
+    ///
+    /// Map the gpio register physical adress to the user space virtual addresses,
+    /// and return the user space virtual adress where it is map.
+    ///
+    /// ## Panics
+    /// if /dev/gpiomem doesn't exist (< Raspbian Jessie), or /dev/gpiomem
+    /// doesn't have the appropriate permissions, or the current user is
+    /// not a member of the gpio group.
+    ///
+    fn map_devgpiomem() -> *mut u32 {
+        // Open /dev/gpiomem with read/write/sync flags. This might fail if
+        // /dev/gpiomem doesn't exist (< Raspbian Jessie), or /dev/gpiomem
+        // doesn't have the appropriate permissions, or the current user is
+        // not a member of the gpio group.
+        let gpiomem_file = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .custom_flags(O_SYNC)
+            .open(PATH_DEV_GPIOMEM)
+            .expect("Problem opening the file /dev/gpiomem");
```

Comment:
maybe use this variable here: `PATH_DEV_GPIOMEM`

Reply:
resolve